### ISSUE TITLE
Setting a gem to a new version range shouldn't fail when setting a specific version in that range passes

### ISSUE THREAD

**Issue by iwz** [rubygems/rubygems#3288]
Depending on your `Gemfile.lock`, when changing `rails` from `~> 5.2` to `~> 6.0`, you may see something similar to the following when running `bundle update rails`:

```
Bundler could not find compatible versions for gem "rails":
  In Gemfile:
    rails (~> 6.0)
    data_migrate was resolved to 6.1.0, which depends on
      rails (>= 5.0)
```

This message is very confusing, since it doesn't seem like there's a conflict there...

However, this goes away if you change `~> 6.0` to `6.0.1`. So, for some reason using a specific version instead of a range is resolving differently. This seems like maybe a UX problem that should be addressed. 

So really the "bug" here is that if you set rails `~> 6.0`, it's only updating the gem you requested (and direct dependencies of that gem which would change), but it doesn't update other things which depend on direct dependencies of that gem. But if you set `rails` to `6.0.1`, it will update those gems even if you didn't ask it to.

Script to reproduce:
https://gist.github.com/iwz/247640c88af7e5bcd962a51654d904f9

**Comment by deivid-rodriguez** [comment]
I verified that this issue still reproduces, and now it also reproduces with an exact version as well.

I minimized it a bit more with the following `Gemfile` and `Gemfile.lock`:

#### Gemfile

```ruby
source "https://rubygems.org"
gem "data_migrate"
gem "rails", "~> 6.0.3.3"
gem "resque-web", "~> 0.0.8"
```

#### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    actioncable (5.2.4)
      actionpack (= 5.2.4)
      nio4r (~> 2.0)
      websocket-driver (>= 0.6.1)
    actionmailer (5.2.4)
      actionpack (= 5.2.4)
      actionview (= 5.2.4)
      activejob (= 5.2.4)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 2.0)
    actionpack (5.2.4)
      actionview (= 5.2.4)
      activesupport (= 5.2.4)
      rack (~> 2.0)
      rack-test (>= 0.6.3)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.2.4)
      activesupport (= 5.2.4)
      builder (~> 3.1)
      erubi (~> 1.4)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.3)
    activejob (5.2.4)
      activesupport (= 5.2.4)
      globalid (>= 0.3.6)
    activemodel (5.2.4)
      activesupport (= 5.2.4)
    activerecord (5.2.4)
      activemodel (= 5.2.4)
      activesupport (= 5.2.4)
      arel (>= 9.0)
    activestorage (5.2.4)
      actionpack (= 5.2.4)
      activerecord (= 5.2.4)
      marcel (~> 0.3.1)
    activesupport (5.2.4)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    arel (9.0.0)
    builder (3.2.3)
    coffee-rails (4.2.2)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.12.2)
    commonjs (0.2.7)
    concurrent-ruby (1.1.5)
    crass (1.0.5)
    data_migrate (6.1.0)
      rails (>= 5.0)
    erubi (1.9.0)
    execjs (2.7.0)
    ffi (1.11.1)
    font-awesome-sass (5.8.1)
      sassc (>= 1.11)
    globalid (0.4.2)
      activesupport (>= 4.2.0)
    i18n (1.7.0)
      concurrent-ruby (~> 1.0)
    jquery-rails (4.3.3)
      rails-dom-testing (>= 1, < 3)
      railties (>= 4.2.0)
      thor (>= 0.14, < 2.0)
    less (2.6.0)
      commonjs (~> 0.2.7)
    less-rails (2.8.0)
      actionpack (>= 4.0)
      less (~> 2.6.0)
      sprockets (> 2, < 4)
      tilt
    loofah (2.4.0)
      crass (~> 1.0.2)
      nokogiri (>= 1.5.9)
    mail (2.7.1)
      mini_mime (>= 0.1.1)
    marcel (0.3.3)
      mimemagic (~> 0.3.2)
    method_source (0.9.2)
    mimemagic (0.3.3)
    mini_mime (1.0.2)
    mini_portile2 (2.4.0)
    minitest (5.13.0)
    mono_logger (1.1.0)
    multi_json (1.14.1)
    mustermann (1.0.3)
    nio4r (2.5.2)
    nokogiri (1.10.7)
      mini_portile2 (~> 2.4.0)
    rack (2.0.7)
    rack-protection (2.0.7)
      rack
    rack-test (1.1.0)
      rack (>= 1.0, < 3)
    rails (5.2.4)
      actioncable (= 5.2.4)
      actionmailer (= 5.2.4)
      actionpack (= 5.2.4)
      actionview (= 5.2.4)
      activejob (= 5.2.4)
      activemodel (= 5.2.4)
      activerecord (= 5.2.4)
      activestorage (= 5.2.4)
      activesupport (= 5.2.4)
      bundler (>= 1.3.0)
      railties (= 5.2.4)
      sprockets-rails (>= 2.0.0)
    rails-dom-testing (2.0.3)
      activesupport (>= 4.2.0)
      nokogiri (>= 1.6)
    rails-html-sanitizer (1.3.0)
      loofah (~> 2.3)
    railties (5.2.4)
      actionpack (= 5.2.4)
      activesupport (= 5.2.4)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.19.0, < 2.0)
    rake (12.3.3)
    redis (4.1.3)
    redis-namespace (1.6.0)
      redis (>= 3.0.4)
    resque (1.27.4)
      mono_logger (~> 1.0)
      multi_json (~> 1.0)
      redis-namespace (~> 1.3)
      sinatra (>= 0.9.2)
      vegas (~> 0.1.2)
    resque-web (0.0.12)
      coffee-rails
      font-awesome-sass
      jquery-rails
      resque
      sass-rails
      twitter-bootstrap-rails
    sass-rails (6.0.0)
      sassc-rails (~> 2.1, >= 2.1.1)
    sassc (2.0.1)
      ffi (~> 1.9)
      rake
    sassc-rails (2.1.1)
      railties (>= 4.0.0)
      sassc (>= 2.0)
      sprockets (> 3.0)
      sprockets-rails
      tilt
    sinatra (2.0.7)
      mustermann (~> 1.0)
      rack (~> 2.0)
      rack-protection (= 2.0.7)
      tilt (~> 2.0)
    sprockets (3.7.2)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (2.3.3)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      sprockets (>= 2.8, < 4.0)
    thor (0.20.3)
    thread_safe (0.3.6)
    tilt (2.0.10)
    twitter-bootstrap-rails (4.0.0)
      actionpack (~> 5.0, >= 5.0.1)
      execjs (~> 2.7)
      less-rails (~> 2.8, >= 2.8.0)
      railties (~> 5.0, >= 5.0.1)
    tzinfo (1.2.5)
      thread_safe (~> 0.1)
    vegas (0.1.11)
      rack (>= 1.0.0)
    websocket-driver (0.7.1)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.4)

PLATFORMS
  ruby

DEPENDENCIES
  data_migrate
  rails (~> 5.2)
  resque-web (~> 0.0.8)
```

It errrors out very fast with:

```
$ bundle lock --update rails 
Fetching gem metadata from https://rubygems.org/.............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "actionpack":
  In Gemfile:
    rails (~> 6.0.3.3) was resolved to 6.0.3.3, which depends on
      actionpack (= 6.0.3.3)

    resque-web (~> 0.0.8) was resolved to 0.0.12, which depends on
      twitter-bootstrap-rails was resolved to 4.0.0, which depends on
        actionpack (~> 5.0, >= 5.0.1)
```

As explained by @iwz, the problem is that `bundler` is only trying to update `rails` and direct dependencies of it. If necessary,  `bundler` should also unlock dependencies introducing incompatible requirements with the Gemfile, in this case, `twitter-bootstrap-rails` should be unlocked so that it can be downgraded to `3.2.2`, which is what a fresh resolution without a lockfile does.

All that said, I'm not sure how to fix this, but I'll keep studying the resolver :nerd_face:.

**Comment by deivid-rodriguez** [comment]
This will be fixed by #5881! 🎉 

With that PR, Bundler will happily upgrade Rails, generating the following diff in the `Gemfile.lock` file

```diff
diff --git a/Gemfile.lock b/Gemfile.lock
index 0b8d8db..c298204 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -1,49 +1,62 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    actioncable (5.2.4)
-      actionpack (= 5.2.4)
+    actioncable (6.0.3.7)
+      actionpack (= 6.0.3.7)
       nio4r (~> 2.0)
       websocket-driver (>= 0.6.1)
-    actionmailer (5.2.4)
-      actionpack (= 5.2.4)
-      actionview (= 5.2.4)
-      activejob (= 5.2.4)
+    actionmailbox (6.0.3.7)
+      actionpack (= 6.0.3.7)
+      activejob (= 6.0.3.7)
+      activerecord (= 6.0.3.7)
+      activestorage (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
+      mail (>= 2.7.1)
+    actionmailer (6.0.3.7)
+      actionpack (= 6.0.3.7)
+      actionview (= 6.0.3.7)
+      activejob (= 6.0.3.7)
       mail (~> 2.5, >= 2.5.4)
       rails-dom-testing (~> 2.0)
-    actionpack (5.2.4)
-      actionview (= 5.2.4)
-      activesupport (= 5.2.4)
-      rack (~> 2.0)
+    actionpack (6.0.3.7)
+      actionview (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
+      rack (~> 2.0, >= 2.0.8)
       rack-test (>= 0.6.3)
       rails-dom-testing (~> 2.0)
-      rails-html-sanitizer (~> 1.0, >= 1.0.2)
-    actionview (5.2.4)
-      activesupport (= 5.2.4)
+      rails-html-sanitizer (~> 1.0, >= 1.2.0)
+    actiontext (6.0.3.7)
+      actionpack (= 6.0.3.7)
+      activerecord (= 6.0.3.7)
+      activestorage (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
+      nokogiri (>= 1.8.5)
+    actionview (6.0.3.7)
+      activesupport (= 6.0.3.7)
       builder (~> 3.1)
       erubi (~> 1.4)
       rails-dom-testing (~> 2.0)
-      rails-html-sanitizer (~> 1.0, >= 1.0.3)
-    activejob (5.2.4)
-      activesupport (= 5.2.4)
+      rails-html-sanitizer (~> 1.1, >= 1.2.0)
+    activejob (6.0.3.7)
+      activesupport (= 6.0.3.7)
       globalid (>= 0.3.6)
-    activemodel (5.2.4)
-      activesupport (= 5.2.4)
-    activerecord (5.2.4)
-      activemodel (= 5.2.4)
-      activesupport (= 5.2.4)
-      arel (>= 9.0)
-    activestorage (5.2.4)
-      actionpack (= 5.2.4)
-      activerecord (= 5.2.4)
-      marcel (~> 0.3.1)
-    activesupport (5.2.4)
+    activemodel (6.0.3.7)
+      activesupport (= 6.0.3.7)
+    activerecord (6.0.3.7)
+      activemodel (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
+    activestorage (6.0.3.7)
+      actionpack (= 6.0.3.7)
+      activejob (= 6.0.3.7)
+      activerecord (= 6.0.3.7)
+      marcel (~> 1.0.0)
+    activesupport (6.0.3.7)
       concurrent-ruby (~> 1.0, >= 1.0.2)
       i18n (>= 0.7, < 2)
       minitest (~> 5.1)
       tzinfo (~> 1.1)
-    arel (9.0.0)
-    builder (3.2.3)
+      zeitwerk (~> 2.2, >= 2.2.2)
+    builder (3.2.4)
     coffee-rails (4.2.2)
       coffee-script (>= 2.2.0)
       railties (>= 4.0.0)
@@ -52,18 +65,19 @@ GEM
       execjs
     coffee-script-source (1.12.2)
     commonjs (0.2.7)
-    concurrent-ruby (1.1.5)
-    crass (1.0.5)
-    data_migrate (6.1.0)
-      rails (>= 5.0)
-    erubi (1.9.0)
+    concurrent-ruby (1.1.10)
+    crass (1.0.6)
+    data_migrate (8.1.1)
+      activerecord (>= 5.0)
+      railties (>= 5.0)
+    erubi (1.11.0)
     execjs (2.7.0)
     ffi (1.11.1)
     font-awesome-sass (5.8.1)
       sassc (>= 1.11)
-    globalid (0.4.2)
-      activesupport (>= 4.2.0)
-    i18n (1.7.0)
+    globalid (1.0.0)
+      activesupport (>= 5.0)
+    i18n (1.12.0)
       concurrent-ruby (~> 1.0)
     jquery-rails (4.3.3)
       rails-dom-testing (>= 1, < 3)
@@ -71,59 +85,60 @@ GEM
       thor (>= 0.14, < 2.0)
     less (2.6.0)
       commonjs (~> 0.2.7)
-    less-rails (2.8.0)
-      actionpack (>= 4.0)
+    less-rails (4.0.0)
+      actionpack (>= 4)
       less (~> 2.6.0)
-      sprockets (> 2, < 4)
-      tilt
-    loofah (2.4.0)
+      sprockets (>= 2)
+    loofah (2.18.0)
       crass (~> 1.0.2)
       nokogiri (>= 1.5.9)
     mail (2.7.1)
       mini_mime (>= 0.1.1)
-    marcel (0.3.3)
-      mimemagic (~> 0.3.2)
-    method_source (0.9.2)
-    mimemagic (0.3.3)
-    mini_mime (1.0.2)
-    mini_portile2 (2.4.0)
-    minitest (5.13.0)
+    marcel (1.0.2)
+    method_source (1.0.0)
+    mini_mime (1.1.2)
+    mini_portile2 (2.8.0)
+    minitest (5.16.3)
     mono_logger (1.1.0)
     multi_json (1.14.1)
     mustermann (1.0.3)
-    nio4r (2.5.2)
-    nokogiri (1.10.7)
-      mini_portile2 (~> 2.4.0)
-    rack (2.0.7)
+    nio4r (2.5.8)
+    nokogiri (1.13.8)
+      mini_portile2 (~> 2.8.0)
+      racc (~> 1.4)
+    racc (1.6.0)
+    rack (2.2.4)
     rack-protection (2.0.7)
       rack
-    rack-test (1.1.0)
-      rack (>= 1.0, < 3)
-    rails (5.2.4)
-      actioncable (= 5.2.4)
-      actionmailer (= 5.2.4)
-      actionpack (= 5.2.4)
-      actionview (= 5.2.4)
-      activejob (= 5.2.4)
-      activemodel (= 5.2.4)
-      activerecord (= 5.2.4)
-      activestorage (= 5.2.4)
-      activesupport (= 5.2.4)
+    rack-test (2.0.2)
+      rack (>= 1.3)
+    rails (6.0.3.7)
+      actioncable (= 6.0.3.7)
+      actionmailbox (= 6.0.3.7)
+      actionmailer (= 6.0.3.7)
+      actionpack (= 6.0.3.7)
+      actiontext (= 6.0.3.7)
+      actionview (= 6.0.3.7)
+      activejob (= 6.0.3.7)
+      activemodel (= 6.0.3.7)
+      activerecord (= 6.0.3.7)
+      activestorage (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
       bundler (>= 1.3.0)
-      railties (= 5.2.4)
+      railties (= 6.0.3.7)
       sprockets-rails (>= 2.0.0)
     rails-dom-testing (2.0.3)
       activesupport (>= 4.2.0)
       nokogiri (>= 1.6)
-    rails-html-sanitizer (1.3.0)
+    rails-html-sanitizer (1.4.3)
       loofah (~> 2.3)
-    railties (5.2.4)
-      actionpack (= 5.2.4)
-      activesupport (= 5.2.4)
+    railties (6.0.3.7)
+      actionpack (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
       method_source
       rake (>= 0.8.7)
-      thor (>= 0.19.0, < 2.0)
-    rake (12.3.3)
+      thor (>= 0.20.3, < 2.0)
+    rake (13.0.6)
     redis (4.1.3)
     redis-namespace (1.6.0)
       redis (>= 3.0.4)
       sprockets-rails (>= 2.0.0)
     rails-dom-testing (2.0.3)
       activesupport (>= 4.2.0)
       nokogiri (>= 1.6)
-    rails-html-sanitizer (1.3.0)
+    rails-html-sanitizer (1.4.3)
       loofah (~> 2.3)
-    railties (5.2.4)
-      actionpack (= 5.2.4)
-      activesupport (= 5.2.4)
+    railties (6.0.3.7)
+      actionpack (= 6.0.3.7)
+      activesupport (= 6.0.3.7)
       method_source
       rake (>= 0.8.7)
-      thor (>= 0.19.0, < 2.0)
-    rake (12.3.3)
+      thor (>= 0.20.3, < 2.0)
+    rake (13.0.6)
     redis (4.1.3)
     redis-namespace (1.6.0)
       redis (>= 3.0.4)
@@ -156,33 +171,37 @@ GEM
       rack (~> 2.0)
       rack-protection (= 2.0.7)
       tilt (~> 2.0)
-    sprockets (3.7.2)
+    sprockets (4.1.1)
       concurrent-ruby (~> 1.0)
       rack (> 1, < 3)
-    sprockets-rails (2.3.3)
-      actionpack (>= 3.0)
-      activesupport (>= 3.0)
-      sprockets (>= 2.8, < 4.0)
-    thor (0.20.3)
+    sprockets-rails (3.4.2)
+      actionpack (>= 5.2)
+      activesupport (>= 5.2)
+      sprockets (>= 3.0.0)
+    thor (1.2.1)
     thread_safe (0.3.6)
     tilt (2.0.10)
-    twitter-bootstrap-rails (4.0.0)
-      actionpack (~> 5.0, >= 5.0.1)
+    twitter-bootstrap-rails (5.0.0)
+      actionpack (>= 5.0, < 8.0)
       execjs (~> 2.7)
-      less-rails (~> 2.8, >= 2.8.0)
-      railties (~> 5.0, >= 5.0.1)
-    tzinfo (1.2.5)
+      less-rails (>= 3.0, < 5.0)
+      railties (>= 5.0, < 8.0)
+    tzinfo (1.2.10)
       thread_safe (~> 0.1)
     vegas (0.1.11)
       rack (>= 1.0.0)
-    websocket-driver (0.7.1)
+    websocket-driver (0.7.5)
       websocket-extensions (>= 0.1.0)
-    websocket-extensions (0.1.4)
+    websocket-extensions (0.1.5)
+    zeitwerk (2.6.0)
 
 PLATFORMS
   ruby
 
 DEPENDENCIES
   data_migrate
-  rails (~> 5.2)
+  rails (~> 6.0.3.3)
   resque-web (~> 0.0.8)
+
+BUNDLED WITH
+   2.4.0.dev
```
